Blazor file download updates (#24560)
dotnet/AspNetCore.Docs: aspnetcore/blazor/file-downloads.md

* Trigger the download via `click`.
* Remove the anchor (`<a>`) element.

> At this point, the file download is triggered, and it's generally safe to revoke the temporary object URL by calling [`revokeObjectURL`](https://developer.mozilla.org/docs/Web/API/URL/revokeObjectURL) on the URL. **This is an important step to ensure memory isn't leaked on the client.**

Reviewer: I like your suggested wording:

`or strike the "generally" language in favor of just saying that the URL should be revoked with the API as indicated.`

As far as I can tell, immediately revoking the object url after starting the download / anchorElement.click() is sufficient as the browser gets a "copy of the object in the blob store" so the revoking the original URL is inconsequential.

https://bugzilla.mozilla.org/show_bug.cgi?id=1282407#c5

Author: ```suggestion
At this point, the file download is triggered and then the temporary object URL is revoked by calling [`revokeObjectURL`](https://developer.mozilla.org/docs/Web/API/URL/revokeObjectURL) on the URL. **This is an important step to ensure memory isn't leaked on the client.**
```

function triggerFileDownload(fileName, url) {

>   const anchorElement = document.createElement('a');

Author: @TanayParikh ... Even without link text, is there any indication in the UI that the anchor exists for large file downloads? If so, can this include ...

```javascript
anchorElement.style.display = 'none';
```

Reviewer 2: Honestly, I've never seen the anchor to show up when added like this

Reviewer: This shouldn't be an issue as we're just creating the element without actually appending it to the document (ex. `document.body.appendChild`). Hence, it shouldn't show up anywhere.

Author: Ok ... thx. I (obviously) haven't had a chance to walk this one (*yet* 😄).

  anchorElement.href = url;
  anchorElement.download = fileName ?? '';
  anchorElement.click();
  anchorElement.remove();
}
```

> `ArrayBuffer` effectively loads the entire file into client memory. For file downloads over 1 GB, we recommend opening a new tab for each large file download by setting the anchor's [`target`](https://developer.mozilla.org/docs/Web/HTML/Element/a#attr-target) to `_blank` and [`referrerpolicy`](https://developer.mozilla.org/docs/Web/HTML/Element/a#attr-referrerpolicy) to `no-referrer`:

Reviewer: ```suggestion
`contentStreamReference.arrayBuffer()` loads the entire file into client memory. For file downloads over 250 MB, we recommend opening a new tab for each large file download by setting the anchor's [`target`](https://developer.mozilla.org/docs/Web/HTML/Element/a#attr-target) to `_blank` and [`referrerpolicy`](https://developer.mozilla.org/docs/Web/HTML/Element/a#attr-referrerpolicy) to `no-referrer`:
```

```javascript
anchorElement.target = '_blank';
anchorElement.referrerpolicy = 'no-referrer';

> ```

Reviewer: I think we may want to consider a more comprehensive example here. While these two lines open the download link in a new tab, we may want to show / mention how we rework the actual download to not use `contentStreamReference.arrayBuffer()`.

Maybe just creating a `triggerFileDownloadInNewTab` would do the trick.

Author: > rework the actual download to not use `contentStreamReference.arrayBuffer()` ... Maybe just creating a `triggerFileDownloadInNewTab` would do the trick.

Yes, could you provide the method for it? ... then I'll draft the text around it and ping u back.

> `contentStreamReference.arrayBuffer()` loads the entire file into client memory. For file downloads over 250 MB, we recommend opening a new tab for each large file download by setting the anchor's [`target`](https://developer.mozilla.org/docs/Web/HTML/Element/a#attr-target) to `_blank` and [`referrerpolicy`](https://developer.mozilla.org/docs/Web/HTML/Element/a#attr-referrerpolicy) to `no-referrer`:
> 
> ```javascript
> anchorElement.target = '_blank';
> anchorElement.referrerpolicy = 'no-referrer';
> ```

Reviewer: ```suggestion
`contentStreamReference.arrayBuffer()` loads the entire file into client memory. For file downloads over 250 MB, we recommend downloading the file from a URL instead:
```

```csharp
<button @onclick="DownloadFileFromURL">
    Download File From Stream
</button>

@code {
    private async Task DownloadFileFromURL()
    {
        var fileURL = "https://aka.ms/dotnet/7.0.1xx/daily/dotnet-sdk-win-x64.exe";
        var fileName = "dotnet-7-nightly";
        await JS.InvokeVoidAsync("triggerFileDownload", fileName, fileURL);
    }
}
```

(feel free to replace the download link with something else, this is just random)

Maybe worth including a blurb about "Please ensure security best practices are followed when providing a URL to download the file" (example text feel free to change as you see fit).

Author: Thanks @TanayParikh ... I'll get this set up in the topic on Tuesday and ping u back. Have a nice holiday weekend! 🌴

Reviewer: Thanks, you too!
